### PR TITLE
Add JSONL caching for dataset.map().

### PR DESCRIPTION
**Pull request by nsthorat**
This means if your map() throws half way through, it will continue the next time, unless overwrite=True.

### CODE REVIEW

File: lilac/data/dataset_duckdb.py [hunk 1]
```diff
@@ -1934,10 +1942,33 @@ def map(
...
     def _rows() -> Iterable[Item]:
-      nonlocal select_queries
+      nonlocal select_queries, con
+
+      # Anti-join removes rows that are already in the cache.
+      anti_join = ''
+      if use_jsonl_cache:
+        anti_join = f"""
+          ANTI JOIN (
+            SELECT * FROM read_json_auto('{jsonl_cache_filepath}', IGNORE_ERRORS=true)) as cache
```
Comment (dsmilkov) [inline lilac/data/dataset_duckdb.py:1966]: add format='newline_delimited' after ignore_errors so duckdb doesn't have to do guessing on jsonl vs json

File: lilac/data/dataset_duckdb.py [hunk 2]
```diff
@@ -23,6 +23,7 @@
 import pandas as pd
 import pyarrow as pa
 import yaml
+from datasets import fingerprint
```
Comment (dsmilkov) [inline lilac/data/dataset_duckdb.py:26]: is this used?

File: lilac/data/dataset_duckdb.py [hunk 3]
```diff
@@ -1993,7 +2042,7 @@ def _rows() -> Iterable[Item]:
     tmp_con.execute(f"""
```
Comment (dsmilkov) [inline lilac/data/dataset_duckdb.py:2042]: do we need to call file.flush() here after we dumps json to disk, in case there is a buffer

**Review by dsmilkov — APPROVED** [review]

File: lilac/data/dataset_duckdb.py (hunk 2 above)
Comment (nsthorat) [reply lilac/data/dataset_duckdb.py]: Nope, autoformatter doesnt remove them anymore?

File: lilac/data/dataset_duckdb.py (hunk 1 above)
Comment (nsthorat) [reply lilac/data/dataset_duckdb.py]: Done

File: lilac/data/dataset_duckdb.py (hunk 3 above)
Comment (nsthorat) [reply lilac/data/dataset_duckdb.py:2044]: Shouldn't need to since it's a context manager

**Pull request merged**